File: Leetcode/linkedlist/817_Linked List Components/Solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def numComponents(self, head, G):
        """
        :type head: ListNode
        :type G: List[int]
        :rtype: int
        """
        s = set()
        for x in G:
            s.add(x)
        res = 0
        while head:
            if head.val in s and (not head.next or not head.next.val  in s):
                res+=1
            head = head.next
        return res;

File: Leetcode/linkedlist/817_Linked List Components/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public int numComponents(ListNode head, int[] G) {
        HashSet<Integer> set = new HashSet<>();
        for(int num: G){
            set.add(num);
        }
        ListNode temp = head;
        int res = 0;
        while(temp!=null){
            if(set.contains(temp.val) && (temp.next==null || !set.contains(temp.next.val)))
                res++;
            temp = temp.next;
        }
        return res;
    }
}
